src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,373 @@
+                //      {
+                //          var x = [|c|].c;
+                //      }
+                //  }
+
+                if (SemanticEquivalence.AreEquivalent(
+                    originalSemanticModel, currentSemanticModel, expressionInOriginal, nodeInCurrent))
+                {
+                    var originalOperation = originalSemanticModel.GetOperation(expressionInOriginal, cancellationToken);
+                    if (IsInstanceMemberReference(originalOperation))
+                    {
+                        var currentOperation = currentSemanticModel.GetOperation(nodeInCurrent, cancellationToken);
+                        return IsInstanceMemberReference(currentOperation);
+                    }
+
+                    return true;
+                }
+            }
+
+            return false;
+            static bool IsInstanceMemberReference(IOperation operation)
+                => operation is IMemberReferenceOperation memberReferenceOperation &&
+                    memberReferenceOperation.Instance?.Kind == OperationKind.InstanceReference;
+        }
+
+        protected TNode Rewrite<TNode>(
+            SemanticDocument originalDocument,
+            TExpressionSyntax expressionInOriginal,
+            TExpressionSyntax variableName,
+            SemanticDocument currentDocument,
+            TNode withinNodeInCurrent,
+            bool allOccurrences,
+            CancellationToken cancellationToken)
+            where TNode : SyntaxNode
+        {
+            var generator = SyntaxGenerator.GetGenerator(originalDocument.Document);
+            var matches = FindMatches(originalDocument, expressionInOriginal, currentDocument, withinNodeInCurrent, allOccurrences, cancellationToken);
+
+            // Parenthesize the variable, and go and replace anything we find with it.
+            // NOTE: we do not want elastic trivia as we want to just replace the existing code 
+            // as is, while preserving the trivia there.  We do not want to update it.
+            var replacement = generator.AddParentheses(variableName, includeElasticTrivia: false)
+                                         .WithAdditionalAnnotations(Formatter.Annotation);
+
+            return RewriteCore(withinNodeInCurrent, replacement, matches);
+        }
+
+        protected abstract TNode RewriteCore<TNode>(
+            TNode node,
+            SyntaxNode replacementNode,
+            ISet<TExpressionSyntax> matches)
+            where TNode : SyntaxNode;
+
+        private class IntroduceParameterCodeAction : AbstractIntroduceParameterCodeAction
+        {
+            internal IntroduceParameterCodeAction(
+                SemanticDocument document,
+                TService service,
+                TExpressionSyntax expression,
+                bool allOccurrences,
+                bool trampoline)
+                : base(document, service, expression, allOccurrences, trampoline)
+            { }
+        }
```

Comment:
i don't understand what's going on here.  why is there an abstract code action if it's just going to have an empty derived tyep that just defers to its constructor?

Reply:
That was the pattern I saw in Introduce Variable as well as Introduce Local For Expression, so I thought that was the accepted design pattern. I will message you more about this